File: 42-trapping-rain-water/trapping-rain-water.py
class Solution(object):
    def trap(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        i, j = 0, len(height)-1
        lh, rh = height[i], height[j]
        total = 0

        while i<j:
            if lh < rh:
                i += 1
                lh = max(lh, height[i])
                total += lh -height[i]
            else:
                j -= 1
                rh = max(rh, height[j])
                total += rh -height[j]
        
        return total


        
        

        

File: 42-trapping-rain-water/trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int i = 0;
        int j = height.length - 1;

        int lh = height[i];
        int rh = height[j];
        int total = 0;

        while(i<j){
            if (lh < rh){
                i++;
                lh = Math.max(lh, height[i]);
                total += (lh - height[i]);
            }
            else{
                j--;
                rh = Math.max(rh, height[j]);
                total += (rh -height[j]); 
            }
        }

        return total;
        
    }
}